tests/Microsoft.Identity.Web.Test/WebAppExtensionsTests.cs

```diff
@@ -1,32 +1,493 @@
+        {
+            var redirectFunc = Substitute.For<Func<RedirectContext, Task>>();
+            var services = new ServiceCollection()
+                .Configure<OpenIdConnectOptions>(_oidcScheme, (options) => {
+                    options.Events ??= new OpenIdConnectEvents();
+                    options.Events.OnRedirectToIdentityProvider += redirectFunc;
+                });
+            services.AddDataProtection();
+
+            if (useServiceCollectionExtension)
+                services.AddSignIn(_configureOidcOptions, _configureMsOptions, _oidcScheme, _cookieScheme, false);
+            else
+                new AuthenticationBuilder(services)
+                    .AddSignIn(_configureOidcOptions, _configureMsOptions, _oidcScheme, _cookieScheme, false);
+
+            await AddSignIn_TestOnRedirectToIdentityProviderEvent(services, redirectFunc).ConfigureAwait(false);
+        }
+
+        private async Task AddSignIn_TestOnRedirectToIdentityProviderEvent(IServiceCollection services, Func<RedirectContext, Task> redirectFunc)
+        {
+            var provider = services.BuildServiceProvider();
+
+            var oidcOptions = provider.GetRequiredService<IOptionsFactory<OpenIdConnectOptions>>().Create(_oidcScheme);
+
+            var httpContext = HttpContextUtilities.CreateHttpContext();
+            var authScheme = new AuthenticationScheme(OpenIdConnectDefaults.AuthenticationScheme, OpenIdConnectDefaults.AuthenticationScheme, typeof(OpenIdConnectHandler));
+            var authProperties = new AuthenticationProperties();
+            authProperties.Items[OidcConstants.AdditionalClaims] = "additional_claims";
+            authProperties.Parameters[OpenIdConnectParameterNames.LoginHint] = "login_hint";
+            authProperties.Parameters[OpenIdConnectParameterNames.DomainHint] = "domain_hint";
+            var redirectContext = new RedirectContext(httpContext, authScheme, oidcOptions, authProperties)
+            {
+                ProtocolMessage = new OpenIdConnectMessage(),
+            };
+
+            await oidcOptions.Events.RedirectToIdentityProvider(redirectContext).ConfigureAwait(false);
+
+            // Assert properties set, events called
+            await redirectFunc.ReceivedWithAnyArgs().Invoke(Arg.Any<RedirectContext>()).ConfigureAwait(false);
+            Assert.NotNull(redirectContext.ProtocolMessage.LoginHint);
+            Assert.NotNull(redirectContext.ProtocolMessage.DomainHint);
+            Assert.NotNull(redirectContext.ProtocolMessage.Parameters[OidcConstants.AdditionalClaims]);
+        }
+
+        [Theory]
+        [InlineData(true, true)]
+        [InlineData(true, false)]
+        [InlineData(false, true)]
+        [InlineData(false, false)]
+        public void AddSignIn_WithConfigName_SubscribesToDiagnostics(bool useServiceCollectionExtension, bool subscribeToDiagnostics)
+        {
```

Comment:
can this be merged w/the AddSignIn_WithConfigActions(bool userServiceCollectionExtension) test, as it's just adding logging, but lots more code to these tests. I think you could streamline it more and do less but more in each test. just a thought. things seem to be repeating with slight changes.

Reply:
Reply in https://github.com/AzureAD/microsoft-identity-web/pull/132#discussion_r417813473